### PR TITLE
Add floating directive

### LINKED ISSUE #2721: [COMPONENT] Multi select
- [x] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the solution you'd like

For use in an ongoing Bankdata project, we need a searchable dropdown, that supports multiple selections.

## Have you considered any alternatives?

Since there will be upwards of 500 items in the dropdown for our use case, it is crucial that the dropdown is searchable.
Additionally, the requirement of selecting multiple values from the dropdown means that the current components are not sufficient.

## Are there any additional context?

Some information from UX is provided below.

### Mock-up
mockup

### Requirements

- [ ] When no items are selected, and the input is not focused, a placeholder text should be shown (Example 01)
- [ ] When the input is focused, the dropdown should start showing (Example 02).
  * This behaviour may have to change slightly in case of async search.
- [ ] The items shown in the dropdown should be filtered as the user types input. (Example 03)
  * Techincal note: A debounce should probably be forced here. UX previously suggested 500 ms, but this is subject to change.
- [ ] The input should show number of selected items when not focused. (Example 06)
- [ ] It should be possible to clear all selections when the input is not focused. (Example 06)
- [ ] When the dropdown is open, selected items should be placed first in the list. (Examples 05 and 07)
- [ ] There should be an option for displaying the total number of items matching a dropdown item. (Example B01)
  * I.e. an outer list with duplicates is filtered based on the selection in the multi select component.
    Each option in the dropdown shows the number of outer list elements with that value.
  * This would probably be achieved by allowing custom item label rendering.
- [ ] It should be possible to mark an item as disabled. (Example B01)
- [ ] It should be possible to mark the entire component as disabled.
- [ ] Provide an async search option, using `Observable` as suggested by @kodeaben [here](https://github.com/kirbydesign/designsystem/issues/2721#issuecomment-1372313120)

#### Single select mode

- [ ] The multi select functionality should be enabled if the `multiple` Input flag is set to true, otherwise the component should function as a searchable dropdown (Only one item can be selected)
- [ ] When in single select mode, the text / label of selected item should be shown instead of the number of selected items
- [ ] When in single select mode, the clear button should still appear when the input is not focused.

### Other

It is not yet clear if scroll virtualization is desirable.

<hr />

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Refinement

- [x] Request that the issue is [UX refined](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#ux-refinement); do not proceed until this is done.
- [x] Request that the issue is [tech refined](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#tech-refinement); do not proceed until this is done.

### Implementation 
The contributor who wants to implement this issue should: 

- [x] Make sure you have read: "[Before you get coding](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#before-you-get-coding)".
- [x] Signal to others you are working on the issue by assigning yourself.
- [ ] Create a branch from the [develop branch](https://github.com/kirbydesign/designsystem/tree/develop) following our [branch naming convention](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Branch). 
- [ ] Publish a WIP implementation to Github as a draft PR and ask for feedback. 
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Update the [cookbook](https://cookbook.kirby.design) with examples and showcases.

### Review
Once the issue has been implemented and is ready for review:

- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review). 
- [ ] Create a pull-request. If you created a draft PR during implementation you can just mark that as "ready for review".

### PR DESCRIPTION
**Pull request by NillerW**
## Which issue does this PR close?

This PR doesn't directly close any issues, but it is needed to close #2721 and #2744 down the road and provide general functionality for popup/floating components. 

The PR Adds a new and better way to handle floating/popup elements and is meant to replace the current popup component. 

This directive is meant for internal use, to create Kirby components such as Dropdown, Tooltip, Popup. etc. and not to be exposed outside the Kirby design system.

## What is the new behavior?

This PR introduces the KirbyFloating directive. This directive can be added directly to HTML-tags to make the tag float relative to another tag. 

The KirbyFloating directive implementation provides the following benefits:
- Ease-of-use: It's simply to attach two attributes to an existing element.
- Reference element and floating element does not have to be placed directly next to each other.
- Not introducing an unnecessary component tag to make the actual content float.
- Trigger events to toggle visibility of the floating element, like click and hover, are handled automatically by the directive.
- Automatic handling of expand direction, based on available size.
- Automatic positioning of floating content based on available size. 

With below image as reference, it can be seen how the new directive works. 

image

Required attributes:
- KirbyFloating: Provides the functionality.
- Reference: Element for which the floating content should be placed next to.

Optional attributes:
- Triggers: How the floating element should be set to show/hide. "Click" will attach a click event to the reference element, which toggles display of the floating element. Supports single value and array.
- Placement: How the element should be placed. Example top, down (under reference element centered), start-down (Under reference element aligned with start of element) etc.

Further inputs to the floating directive is likely to come, based on use-cases and requirements.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Review  
- [x] Determine if your changes are a fix, feature or breaking-change, and add the matching label to your PR. If it is tooling, dependency updates or similar, add ignore-for-release.
- [x] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [x] Request that the changes are code-reviewed 
- [x] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be merged to `develop` by Team Kirby.

### CODE REVIEW

File: apps/cookbook/src/app/examples/dropdown-example/examples/popover.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { Component, Input } from '@angular/core';
```
Comment (RasmusKjeldgaard) [inline apps/cookbook/src/app/examples/dropdown-example/examples/popover.ts]: Seems unrelated, consider removing as the dropdown popover functionality should be replaced with this floating utility altogether pretty soon.

File: apps/cookbook/src/app/showcase/dropdown-showcase/dropdown-showcase.component.html [hunk 2]
```diff
@@ -12,6 +12,15 @@ <h3>Default</h3>
...
 </cookbook-example-viewer>
+<h3>Popup</h3>
+<cookbook-example-viewer [html]="popupExample.template">
+  <div>
+    <cookbook-dropdown-example-popover
+      [size]="size"
+      #popupExample
+    ></cookbook-dropdown-example-popover>
+  </div>
+</cookbook-example-viewer>
```
Comment (RasmusKjeldgaard) [inline apps/cookbook/src/app/showcase/dropdown-showcase/dropdown-showcase.component.html]: Seems unrelated, consider removing as the dropdown popover functionality should be replaced with this floating utility altogether pretty soon.

File: libs/designsystem/shared/floating/src/floating.directive.ts [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+/**
+ * @summary FloatingDirective is a utility that lets you declarative anchor "popup" containers to another element.
+ *
+ * Uses floating-ui, with this directive wraps the functionality: https://floating-ui.com/docs/getting-started
+ *
+ * Uses PortalDirective to enable functionality for portaling the floated content. This should be used when needed
+ * and not as the default option.
+ *
+ * @status In development
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.ts]: Around here, everything is in development 😛 Consider removing.

File: libs/designsystem/shared/floating/src/floating.directive.ts [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+import { from } from 'rxjs';
+import { PortalDirective } from '@kirbydesign/designsystem/shared/portal';
+
+export type TriggerEvent = 'hover' | 'click' | 'focus';
+
+export enum FloatingOffset {
+  none = 0,
+  small = 5,
+  medium = 10,
+  large = 15,
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.ts]: Have these values been aligned with designers?
To me it looks like our standard offset for the dropdown is currently 8px, but that needs to be verified.

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+  @ViewChild(FloatingDirective, { static: true }) public floatingDirective: FloatingDirective;
+
+  /** Start values set to match directive */
+  @Input() public isDisabled: boolean = false;
+  @Input() public closeOnEscapeKey: boolean = true;
+  @Input() public closeOnSelect: boolean = true;
+  @Input() public closeOnBackdrop: boolean = true;
+  @Input() public strategy: Strategy = 'absolute';
+  @Input() public placement: Placement = 'bottom-start';
+}
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.spec.ts:45]: Cant we just not initialise them, so is uses the default values from the directive, and then you set the input when needed?

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+  it('should be created', () => {
+    expect(spectator.component).toBeTruthy();
+  });
+
+  describe('input', () => {
+    describe('reference', () => {
+      it('should set', () => {
+        directive.reference = component.floatingElementRef;
+        expect(directive.reference).toEqual(component.floatingElementRef);
+      });
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.spec.ts:73]: This test seems redundant, as we forcibly assign the value on the first line, before checking what we just did. 

What do we actually want to test here?

Comment (RasmusKjeldgaard) [reply libs/designsystem/shared/floating/src/floating.directive.spec.ts:73]: Unless there is a really good reason, I would not test the getter/setter functionality of Angular. The code for setting the local variable inside the component is very simple, and it is hard to test anything meaningful. 

This goes for all the 'should set' tests in my opinion.

File: libs/designsystem/shared/floating/src/floating.directive.ts [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+interface EventMethods {
+  event: string;
+  method: () => void;
+}
+
+type EventListener = () => void;
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.ts]: The Typescript dom lib has a similar type, so you should be able to remove this line.

File: libs/designsystem/shared/floating/src/floating.directive.ts [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    }
+  }
+
+  private tearDownEventHandling(): void {
+    this.eventListeners.forEach((eventListener) => {
+      if (eventListener != null) {
+        eventListener();
+      }
+    });
+    this.eventListeners = [];
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.ts]: For performance reasons, to not risk collecting a lot of detached elements, shouldnt we actually remove any attached event-listeners here instead?

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+        autoUpdateFuncSpy = jasmine.createSpy('autoUpdate').and.returnValue(null);
+        spyOnProperty(floatingUi, 'autoUpdate', 'get').and.returnValue(autoUpdateFuncSpy);
+
+        computePositionFuncSpy = jasmine
+          .createSpy('computePosition')
+          .and.returnValue(new Promise(null));
+        spyOnProperty(floatingUi, 'computePosition', 'get').and.returnValue(computePositionFuncSpy);
+      });
+
+      describe('add styling', () => {
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/shared/floating/src/floating.directive.spec.ts]: I believe this describe is redundant, and could be removed for readablility so it reads:
`FloatingDirective method ngOnInit should add style XYZ` instead of `FloatingDirective method ngOnInit add style should add style XYZ`

**Review by RasmusKjeldgaard — CHANGES_REQUESTED** [review]
I think this looks great! Only some small comments that should not take long to resolve.

We need to see if we can clean up all the unrelated ng-package changes (consider checking each one out from develop and removing the pre-commit hook in package.json on that commit).

File: libs/designsystem/shared/floating/src/floating.directive.ts (hunk 4 above)
Comment (NillerW) [reply libs/designsystem/shared/floating/src/floating.directive.ts]: I'll check in with Henrik :)

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts (hunk 5 above)
Comment (NillerW) [reply libs/designsystem/shared/floating/src/floating.directive.spec.ts:45]: Since I bind them to the directive, setting them to undefined would also give that value to the directive, thus not being able to use the default values. 

I need these inputs to be able to simulate actually changing the values from the component using the directive :)

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts (hunk 6 above)
Comment (NillerW) [reply libs/designsystem/shared/floating/src/floating.directive.spec.ts:73]: Generally I agree, I just thought for a component library, it seemed reasonable to be sure. 

Especially for reference, since the code is more than a simply setter:

`@Input() public set reference(ref: ElementRef) {
    this.tearDownEventHandling();
    this._reference = ref as ElementRef;
    this.setupEventHandling();
    this.autoUpdatePosition();
  }`
  
  So I'm making sure that if someone changes the affected code, we still get the right reference set.

File: libs/designsystem/shared/floating/src/floating.directive.ts (hunk 8 above)
Comment (NillerW) [reply libs/designsystem/shared/floating/src/floating.directive.ts]: Can you elaborate. The if statement should not be necessary, but just in-case so we are sure to be able to execute the function

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts (hunk 5 above)
Comment (RasmusKjeldgaard) [reply libs/designsystem/shared/floating/src/floating.directive.spec.ts:45]: I see 👍

File: libs/designsystem/shared/floating/src/floating.directive.ts (hunk 8 above)
Comment (RasmusKjeldgaard) [reply libs/designsystem/shared/floating/src/floating.directive.ts]: It dawns on me that I misread this, I was not aware that this utilizes the teardown function returned from renderer. 

But that hightlights the actual problem here, I believe: The naming is a bit misleading, because the value returned and pushed to the eventListeners array is not actually an event listener, but rather an `EventListenerDisposeFunction` or whatever we might call that - so my comment from before about removing the EventListener type is actually wrong, it should be replaced with something like that instead. And then we should rename the array to reflect that.

File: libs/designsystem/shared/floating/src/floating.directive.spec.ts (hunk 6 above)
Comment (RasmusKjeldgaard) [reply libs/designsystem/shared/floating/src/floating.directive.spec.ts:73]: Okay, fair point!

**Review by RasmusKjeldgaard — APPROVED** [review]
🎉
